NAT symmetric logic, eligible users as relays, successfull connections logs, fixed false server responses

AlexDuna opened the pull request:
# NAT-symmetric support + cleaning client behavior

## 1) I kind of transformed our server into a TURN-like(Travel Using Relays around NAT) server for symmetric NATs
When two peers can't punch through, we need the server to relay traffic.
Changes I've made:
- **/src/signaling/handlers/request_relay.rs**
  - **REQUEST_RELAY <server_id> <channel> <username>** marks the user in Channel.need_server_relay
  - **handle_data_from_client()** validates sender and forwards **DATA** to all other users (server acts as relay)
- **/src/client/bin/client.rs**
  - added new flag **send_via_server** - when I must send via server (I'm symmetric or lone), and renamed the old **server_relay_enabled** flag to **channel_has_server_relays** (when I'm relay, mirror what I forward to server as well, because some peers are server-relayed), to make more sense
  - in **process_server_response()**
    - On **MODE SERVER_RELAY <me>** -> sets send_via_server = true ("I will send via server from now on")
    - On **MODE SERVER_RELAY <other>** and if I'm relay -> sets channel_has_server_relays = true ("Relay will also mirror traffic to server")
- **/src/client/networking.rs**
  - **hole_punching_loop()**
    - now uses the new flag **send_via_server**: it sends hole punch only to eligible peers, but peers marked with use_server_relay true, or already connected users are skipped. This way, the hole punches stop when connection with Symmetric NAT users is impossible to achieve.
    - i also added a backoff, which is a HashMap<String, u64> (key = username, value = delay in milliseconds). For every punching try for a peer, the backoff is doubled: initially 500ms, then 1000ms, then 2000ms, at max of 3000ms.
    - Why is it useful? Well, i don't want to send HOLE_PUNCH very very often, as the protocol used to do, it wastes traffic. This way we send less punches after the first tries, we save resources and the logs don't explode. Basically, the punches gradually decrease.
    - First tries are still fast (500ms -> 1s), and we make them more rare as time passes.
    - It helps with my logic of **REQUEST_RELAY** and **MODE SERVER_RELAY**, because after a few seconds the connection is not established, it goes into the **REQUEST RELAY** logic.

## 2) Talking about eligible peers :)
- Because we now have different logics to handle different NAT types, of course i needed to add a check for eligibility of a user.
- What i mean by this is, that i added in **/src/signaling/notifications.rs** and **/src/signaling/heartbeat.rs** the function **pick_eligible_relay**, that checks if a user is on Symmetric NAT (basically if it needs SERVER RELAY), or not.
- We need this check because, in many functions, the promotion of a user relay was made by simply picking the first user in a channel user list to be relay. Well, we don't want a Symmetric NAT user to act as relay for the others, because it would make no sense. All the traffic would need to pass through the server. SO, i check every time a user has to be promoted to relay, or the relay left and we need a new one, and in all kinds of scenarios, if the next relay, would be eligible or not!

## 3) We now have a clear "peer connected" log, when a peer establishes connection
I've made this change so we know every time a DIRECT or SERVER-RELAYED connection is established.
- in **/src/client/handlers.rs**
  - added a one-time print everytime a connection is established, and mark it as **(direct)** or **(via server)**
  - so now after a normal user is connected we should see in the logs something like: "Peer X is now CONNECTED (direct) at ...", and when a Symmetric user is connected: "Peer X is now CONNECTED (via server) ..."

## 4) No more weird "Server response: PING/HOLE_PUNCH"
During setup, peer packets could be misread as "server responses", that would print weird logs.
So i added this logic: in **/src/bin/client.rs**, the signaling endpoint is now a concrete SocketAddr, so we now classify packets by comparing **socket addresses**.
- So, if **src == server_socketaddr** -> treat as server, else treat as peer, so no more "Server response: " prints from peers packets
One more change that helps with this, is that if a server packet contains **"MODE ..."** line, we exit early from setup, so we immediately switch to the main loop.

## 5) Added the necessary fields inside PeerInfo structure

AlexDuna committed:
NAT symmetric support + cleaner client behavior

AlexDuna committed:
Changes regarding choosing eligible user as relay.

kickhead13 commented on src/client/handlers.rs:
```diff
@@ -135,7 +161,7 @@ pub fn handle_data_message(
     peers: &Arc<Mutex<Vec<PeerInfo>>>,
     is_relay: &Arc<Mutex<bool>>,
-    server_relay_enabled: &Arc<Mutex<bool>>,
+    channel_has_server_relays: &Arc<Mutex<bool>>,
```
Is it really necessary for this to be wrapped into Arch<Mutex> ? If it is set only once it can simply be a Mutex<> or maybe Rc<Mutex<>> ?

kickhead13 commented on src/signaling/heartbeat.rs:
```diff
@@ -1,41 +1,81 @@
+    channel
+        .users
+        .iter()
+        .find(|u| !channel.need_server_relay.contains(&u.name))
```
Kind of a weird idea to let the channel use information about the state of the user...

Why does the channel have a field like 'need_server_relay'? The channel *kind of* needs server relay but it makes more sense to say that the user "needs_server_relay". No?

So for me it would seem simpler for the User struct to have a field like "needs_server_relay" instead of the channel.

This would also improve this bit of code, because here we iterate through the users and then to check if any user needs server relay we iterate through channel members... If we had "needs_server_relay" as a User attribute we could simply iterate throuh the users (as we already do) without the need to also iterate through need_server_relay.contains().

Comparison:
Current implementation:
```
.find(|u| !channel.need_server_relay.contains(&u.name)
``` 
My proposal:
```
.find(|u| !u.need_server_relay)
```

kickhead13 commented on src/signaling/handlers/request_relay.rs:
```diff
@@ -36,23 +36,58 @@ pub async fn handle_data_from_client(
+
+                //if sender is relay -> deliver only to need_server_relay
+                if relay_name.as_deref() == Some(sender) {
+                    if channel.need_server_relay.is_empty() {
```
This could be instead a bool, (or Arc<Mutex<bool>> etc.). That gets set whenever a new user joins. Because come to think of it this would be a list of users... no? 

So channel has 2 lists of users (User Structs > bool):
- users
- need_server_relay (as Strings)

So we could have something like:
```
struct User{
   ...
   pub needs_server_relay: bool
}

struct Channel {
 pub users: Vec<User>
 ...
 pub need_server_relay: bool
}
...
impl Channel {
    pub fn add_user(self: &mut Self, user: User) -> () {
        self.channel.users.push(user);
         if user.needs_server_relay {
            self.channel.need_server_relay = true;
        }
    }
}
```

kickhead13 commented on src/client/structures.rs:
```diff
@@ -7,6 +7,7 @@ pub struct PeerInfo {
     pub connected: bool,
-
     pub created_at: Instant,
+    pub use_server_relay: bool, //server will carry traffic for this peer
```
Hmm something like this...

kickhead13 commented on src/client/networking.rs:
```diff
@@ -31,28 +32,50 @@ pub fn start_heartbeat(
                 }
+
+                let entry = backoff.entry(p.username.clone()).or_insert(500);
+                *entry = (*entry).saturating_mul(2).min(3000);
```
Wow cool function didn't know saturating_mul exists. Cool

kickhead13 commented on src/client/handlers.rs:
```diff
@@ -157,8 +193,16 @@ pub fn handle_data_message(
 pub fn handle_peer_message(peers: &Arc<Mutex<Vec<PeerInfo>>>, src: std::net::SocketAddr) {
     let mut guard = peers.lock().unwrap();
     if let Some(p) = guard.iter_mut().find(|p| p.addr == src) {
+        let was_connected = p.connected;
```
Why do we do this? Can we not just put the if & print before setting p.connect to true? I don't believe this variable is necessary.

I believe this should work fine:
```
if !p.connected {
    println!(
            "Peer {} is now CONNECTED (direct) at {}.",
            p.username, p.addr
    );
}
p.last_pong = Instant::now();
p.connected = true;
```

kickhead13 replied:
This also (maybe)

kickhead13 commented on src/client/networking.rs:
```diff
@@ -31,28 +32,50 @@ pub fn start_heartbeat(
     loop {
+        if *send_via_server.lock().unwrap() {
+            // i'm symmetric; i'll use server, no need to punch
+            thread::sleep(Duration::from_millis(500));
```
Why 500 milliseconds? Is this a magic number? Can we make a static variable for it if it is?

```
pub static SERVER_RELAY_SLEEP_MILLISECONDS: u32 = 500;

...

thread::sleep(Duration::from_millis(SERVER_RELAY_SLEEP_MILLISECONDS));
```

kickhead13 commented on src/client/networking.rs:
```diff
@@ -31,28 +32,50 @@ pub fn start_heartbeat(
         }
-        thread::sleep(Duration::from_millis(500));
+
+        let sleep_ms = backoff.values().min().copied().unwrap_or(500);
```
This 500 also seems like a "magic number"... Would be cool to also have it as a static.

kickhead13 replied:
Come to think of it (as we discussed) it would be better to use thread "Conditions" here:
https://doc.rust-lang.org/std/sync/struct.Condvar.html

kickhead13 reviewed (changes requested):
Please look into these magic numbers and that one was_connected variable.

AlexDuna committed:
Condvar for threads, magic numbers to constants, changed flags to AtomicBool.

AlexDuna commented:
After going through your suggestions, i managed to implement the following to come up with fixes:

## 1.Created a new function ensure_connected 
- in /client/handlers.rs
- to be more organized, because otherwise i would just copy the same code in 3 different areas
- this function does what reviewer kickhead13 suggested, instead of adding a new variable to be able to print a CONNECTION message, i could just simply check with a simple if statement peer connection status, print and mark peer connected

## 2. Changed "channel_has_server_relays" and "send_via_server" flags from "Arc<Mutex<bool>>" to "Arc AtomicBool" 
- in /bin/client.rs and /client/handlers.rs
- a simple Mutex<Bool> that you suggested i don't think is enough by itself. To share the same flag instance between multiple threads (main loop, relay keepalive, user input, hole punching), we need **reference-counted ownership** -> **Arc<>**
- here, the flag *channel_has_server_relays* can flip **after** the threads are running, for example after server later announces **MODE SERVER_RELAY**, so it's a concurrent write - concurrent reads situation
- BUT, i found some cleaner solution to this suggestion of yours, that's somehow in between of my implementation and your suggestion: [Arc<AtomicBool>][https://doc.rust-lang.org/std/sync/atomic/struct.AtomicBool.html]
- AtomicBool is a boolean type that can be safely shared between threads. It has the same size, alignment and bit validity as **bool**.
- by using AtomicBool, **reading** and **writing** are **lock-free**, which is perfect for **set once, read many** situation
- maybe for a better understanding of what **Ordering** is, you can checkout this link: [ordering_explained][https://stackoverflow.com/questions/30407121/which-stdsyncatomicordering-to-use]
- basically, i only used Ordering::Acquire and Ordering::Release, a simple example would be:
  - ```store(true, Ordering::Release)``` would mean: thread A: "done, i've set this flag to true, and everything that i've done until this step is now visible"
  - ```load(Ordering::Acquire)``` would mean: thread B: "i'm reading this value, and everything that comes after this reading is not allowed to execute before it".
  - this is just a safe way of reading a flag set by another thread

## 3. Changed all magic numbers to static variables (added a few constants) 
- in /bin/client.rs and /client/networking.rs

## 4. User structure now has a new field that is getting set when a specific user needs server relay
- changes in: 
    - /signaling/structures.rs
    - /signaling/handlers/request_relay.rs
    - /signaling/heartbeat.rs
    - /signaling/handlers/notifications.rs
- now, instead of keeping a list of users that need server relay for every channel, we simply just verify per user if it needs server relay (if symmetric or other strange reason for no ability to connect), so the channel doesn't need to know information about the state of a user.

## 5. I added "Condition variables" to have the ability to block threads such that it consumes no CPU time while waiting for events to occur. 
- Changes in: 
    - /client/structures.rs, 
    - /bin/client.rs, 
    - /client/handlers.rs, 
    - /client/networking.rs 
- condition variables were added for hole_punching threads, but also for relay_keepalive threads, so now there is no CPU consumption while a thread waits to become active.
- this change eliminated useless polling and now every state change is instant, no need to wait a specific amount of seconds until a thread becomes active for a job. Through condvar, threads know exactly when to start working
- in my opinion, it works great with AtomicBool (where i managed to implement this), because AtomicBool offers low latency and big concurrency, and condvar offers everything discribed above.

kickhead13 commented on src/bin/client.rs:
```diff
@@ -129,12 +172,13 @@ fn handle_recv_result(
                     is_relay,
                     channel_has_server_relays,
                     signaling_addr,
+                    punch_sync,
                 );
             }
             true
         }
         Err(e) if e.kind() == std::io::ErrorKind::WouldBlock => {
-            thread::sleep(Duration::from_millis(20));
+            thread::sleep(Duration::from_millis(SETUP_POLL_SLEEP_MS));
             true
```
Why do we wait here? Are we waiting for a specific "condition"?

The point of this "magic number" problem is not only setting them up in static consts (which is good), but also to evaluate whether they are actually needed.

kickhead13 commented on src/bin/client.rs:
```diff
@@ -149,13 +193,15 @@ fn server_responses_during_setup(
 ) {
     let mut buf = [0u8; 1024];
-    let setup_deadline = Instant::now() + Duration::from_millis(1500);
+    let setup_deadline = Instant::now() + Duration::from_millis(SETUP_DEADLINE_MS);
     let mut saw_mode = false;
 
     while Instant::now() < setup_deadline && !saw_mode {
```
I'm not sure why we do this.

kickhead13 reviewed (changes requested):
These need to be explained.

AlexDuna commented:
> These need to be explained.

```const SETUP_DEADLINE_MS```
- this line doesn't wait, it sets a time for the setup phase. It's just a time constant so the client doesn't sit in "setup" phase forever if the server never replies.
- this is totally fine to keep and i don't think we need to remove it


```
const SETUP_POLL_SLEEP_MS: u64 = 20;
const MAIN_POLL_SLEEP_MS: u64 = 50;
````
- those sleeps run on the **WouldBlock** branch for the non-blocking socket.
- In the ```handle_recv_result()```: this is called during the short setup window (until saw_mode == true or the SETUP_DEADLINE_MS passes). The socket is non-blocking
    - if there is a packet ready that contains "MODE ...", saw_mode gets set to true and continues, no sleeping here.
    - on the error branch, if the socket had no packet available at that moment, because its non-blocking, recv_from returns with WouldBlock instead of waiting.
    - the line ```thread::sleep(Duration::from_millis(SETUP_POLL_SLEEP_MS));``` puts the thread to sleep while there is no data available. It wakes after ```SETUP_POLL_SLEEP_MS``` ms (20) and checks again

- In ```main_loop()```: pretty much the same logic as above

- the problem with this sleep method is that if any packet arrives right after the thread goes to sleep, it won't see it until it wakes up again, after the number of ms in SETUP_POLL_SLEEP_MS or MAIN_POLL_SLEEP_MS

- it prevents a busy wait, where the thread repeatedly checks a condition without ever blocking, so it keeps the CPU doing useless work when there's no traffic

After you pointed this out, i really don't like this current approach because it wastes time, CPU usage and creates a little bit of latency, but also it was kind of hard for me to find another method. I have 2 proposals:
1) Make the UDP socket **blocking** and add read timeout. Even tho i've read a lot that in practice we almost always want to use non-blocking, i believe that in this case, because i only use 1 socket here, making it **blocking** would work.
    - we would remove the ```WouldBlock``` error branch and the ```thread::sleep```, the loop would call recv_from, and it would block until:
        - a packet arrives or a timeout consumed, which would mean something like "wake up, even tho no data arrived", need to do this to keep the heartbeats and all
        - here the timeout would not be a magic number, but the loop's heartbeat
2) Event loop with [mio](https://docs.rs/mio/latest/mio/)
    - we would keep the socket non-blocking
    - instead of calling ```recv_from()``` in a loop and sleeping when there's nothing to read, mio asks the OS when the socket is ready
    - the OS wakes it only when there's some event happening (data to read). No busy-wait, no sleeps
    - the advantage here is that this method is the fastest i found and it can manage multiple sockets, even tho not needed here yet, but better for the future and better than **blocking** method
    - this solution for me seems the best. Harder to implement, but i think it would be for the best.

Waiting on your opinion.

kickhead13 commented:
The mio/polling idea seems to probably be the one to choose.

Let's make this a different PR tho. Maybe create a ticket for this in the backlog.

We will merge this PR, and we'll see when we do this improvement in the future.

kickhead13 reviewed (approved):
Ready for merge.
